src/Raven.Server/Documents/Queries/Results/QueryResultRetrieverBase.cs

```diff
@@ -116,25 +175,88 @@ protected void FinishDocumentSetup(Document doc, Lucene.Net.Search.ScoreDoc scor
 
         public abstract bool TryGetKeyLucene(ref RetrieverInput retrieverInput, out string key);
 
         public abstract bool TryGetKeyCorax(IndexSearcher searcher, long id, out UnmanagedSpan key);
 
-        public abstract Document DirectGet(ref RetrieverInput retrieverInput, string id, DocumentFields fields);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public virtual TDocument DirectGet(ref RetrieverInput retrieverInput, string id, DocumentFields fields)
+        {
+            return DirectGetInternal(ref retrieverInput, id, fields);
+        }
+
+        public virtual void ClearCache()
+        {
+            _loadedDocumentCache.Clear();
+        }
+
+        private TDocument DirectGetInternal(ref RetrieverInput retrieverInput, string id, DocumentFields fields)
+        {
+            if (_loadedDocumentCache.TryGetValue(id, out var doc) == false)
+            {
+                doc = _loadedDocumentCache[id] = DocumentsStorage.Get<TDocument>(DocumentContext, id, fields);
+            }
+
+            // When we take document from cache we've to increase the reference counter.
+            if (typeof(TDocument) == typeof(QueriedDocument))
+                ((QueriedDocument)(object)doc)?.IncreaseReference();
+
+            return doc;
+        }
 
-        protected abstract Document LoadDocument(string id);
+        protected virtual TDocument LoadDocument(TDocument parentDocument, string id, ref RetrieverInput retrieverInput)
+        {
+            using var _ = (_loadScope = _loadScope?.Start() ?? _projectionScope?.For(nameof(QueryTimingsScope.Names.Load)));
+            if (id == string.Empty) // string.Empty means it's original document.
+                return parentDocument;
+            
+            if (_loadedDocumentCache.TryGetValue(id, out var doc))
+            {
+                if (typeof(TDocument) == typeof(QueriedDocument))
+                {
+                    var queriedDocument = (QueriedDocument)(object)doc;
+                    queriedDocument?.IncreaseReference();
+                    
+                    var parent = (QueriedDocument)(object)parentDocument;
+                    parent.LinkReferencedDocument(queriedDocument);
+                    
+                }
+
+                if (typeof(TDocument) != typeof(Document))
+                    return doc;
+            }
+
+            doc = DirectGetInternal(ref retrieverInput, id, DocumentFields.All);
+
+            if (doc != null)
+            {
+                _loadedDocumentCache[id] = doc;
+
+                if (typeof(TDocument) == typeof(QueriedDocument))
+                {
+                    var queriedDocument = (QueriedDocument)(object)doc;
+                    queriedDocument?.IncreaseReference();
+                    
+                    var parent = (QueriedDocument)(object)parentDocument;
+                    parent.LinkReferencedDocument(queriedDocument);
+                }
+            }
```

Comment:
I wonder if we can do something with those checks:

```
if (typeof(TDocument) == typeof(QueriedDocument))
```

Let's talk directly about that.